modules/ui/maproulette_editor.js

```diff
@@ -185,49 +198,32 @@ export function uiMapRouletteEditor(context) {
       .append('div')
       .attr('class', 'detail-section')
       .merge(detailSection);
-
     const osm = context.services.osm;
     if (!osm) return;
 
-    // Add warning if user is not logged in
-    const hasAuth = osm.authenticated();
+    // Check if user is authenticated with OSM
+    const osmAuth = osm.authenticated();
     let authWarning = detailSection.selectAll('.auth-warning')
-      .data(hasAuth ? [] : [0]);
-
-    authWarning.exit()
-      .transition()
-      .duration(200)
-      .style('opacity', 0)
-      .remove();
-
-    let authEnter = authWarning.enter()
-      .insert('div', '.tag-reference-body')
-      .attr('class', 'field-warning auth-warning')
-      .style('opacity', 0);
-
-    authEnter
-      .call(uiIcon('#rapid-icon-alert', 'inline'));
-
-    authEnter
-      .append('span')
-      .text(l10n.t('map_data.layers.maproulette.login'));
-
-    authEnter
-      .append('a')
-      .attr('target', '_blank')
-      .call(uiIcon('#rapid-icon-out-link', 'inline'))
-      .append('span')
-      .text(l10n.t('login'))
-      .on('click.note-login', d3_event => {
-        d3_event.preventDefault();
-        osm.authenticate();
-      });
-
-    authEnter
-      .transition()
-      .duration(200)
-      .style('opacity', 1);
-
+      .data(osmAuth ? [] : [0]);
+    // If not authenticated with OSM, show OSM login prompt
+    if (!osmAuth) {
+      showAuthWarning(authWarning, 'map_data.layers.maproulette.login');
+    }
+    // Check if _mapRouletteApiKey is defined
+    getMapRouletteApiKey(context, (err, apiKey) => {
+      if (err) {
+        console.error(err); // eslint-disable-line no-console
+        return;
+      }
+      _mapRouletteApiKey = apiKey;
+      const mrAuth = _mapRouletteApiKey !== undefined;
+      authWarning = detailSection.selectAll('.auth-warning')
+        .data(mrAuth ? [] : [0]);
+      // If _mapRouletteApiKey is not defined, show MapRoulette login prompt
+      if (!mrAuth) {
+        showAuthWarning(authWarning, 'map_data.layers.maproulette.loginMaproulette');
```

Comment:
Same thing here - I'm sure it works when you test it , but this jumps out at me as unusual.

It's really weird to repurpose the `authWarning` selection, and then only run some of the code after - this is like not giving d3 the opportunity to do its job and update the DOM properly.

A better way to think about d3 is - I might want to show _some_ warnings here, so I put those into the data array, and let d3 handle the work of updating the DOM for me.

```js
// always start by preparing an array of data set up 
// exactly how you want it, everything flows from there
const warnings = [];
if (!osmAuth) warnings.push('hey you better login to osm');
if (!mrAuth) warnings.push('hey you better login to maproulette');

let warnings = selection.selectAll('.warning')
  .data(warnings, d => d);   

// `d => d`  means, the data _is_ the text, there is no key that identifies it.
// In this case, we could leave it off, d3 will use the array index as the key.
//   (and for an array of 2 things I just made, it would be fine).. 
// But for more complicated data it's important to provide a key lookup function.

warnings.exit()   // remove the warnings that should not be shown
  .transition(…)
  .remove();

warnings.enter()   // add the warnings that should be shown
  .append('div')
  .attr('class', 'warning')
  .text(d => d);

// if `warnings` data could change, we'd write this in a way that we 
// get an update selection here and respond to the change in the data
```